### ISSUE TITLE
Squid3 v1.1.6: Perm1 Th-centric expression for U_Concen_RM is mismatched

### ISSUE THREAD

**Issue by sbodorkos** [CIRDLES/Squid#311]
Another little one, and one which can be worked around in the Expression Manager once you figure out what's going on, but which should be fixed in the code:

I am running a Perm1-style Task in Th-centric mode, with directives and "four horsemen" as below:

![image](https://user-images.githubusercontent.com/17473203/56169063-5737b780-6020-11e9-8929-19458c91b5cc.png)

This entails calculation of U (ppm) values in "secondary parent element" mode (i.e. analogous to how we usually calculate Th (ppm), most of the time). But the Built-in U_Concen_RM expression was flagged as Mismatched, and when I checked it, it read:

Th_Concen_RM / 232Th238U * L1033

and of course it **should** read:

Th_Concen_RM / **232Th238U_RM** * L1033

I think this is just a hangover from the days where RefMat 232Th/238U was labelled simply "232Th238U" (remembering that Unknown 232Th/238U was "232Th238US" at that time). Now that we have reversed the naming convention ("232Th238U_RM" and "232Th238U" respectively), it is easy to mismap an expression-name that has always been valid, but whose definition has changed!

**Comment by bowring** [comment]
It turns out that even with this fix, there is a problem that involves circular logic and that derives from discussions at the end in issue #164 relative to Thorium  that I guess I don't fully understand.  We never tested the Thorium scenario until now.

If Uranium is the concentration of interest and directALTPD is false, we use this code to define the expressions:
```java
       if (!isDirectAltPD) {
           Expression expressionPpmTh = buildExpression(TH_CONCEN_PPM,
                   U_CONCEN_PPM * TH_U_EXP / L1033, false, true, false);

           Expression expressionPpmURM = buildExpression(U_CONCEN_PPM_RM,
                   PARENT_ELEMENT_CONC_CONST / AV_PARENT_ELEMENT_CONC_CONST * REF_U_CONC_PPM, true, false, false);

           Expression expressionPpmTh_RM = buildExpression(TH_CONCEN_PPM_RM,
                   U_CONCEN_PPM_RM * TH_U_EXP_RM /  L1033, true, false, false);
            }
```

BUT, if Thorium is the concentration of interest and directALTPD is false, we use this code (which has the circular logic within the second and third expressions):
```java
      if (!isDirectAltPD) {
           Expression expressionPpmU = buildExpression(U_CONCEN_PPM,
                  TH_CONCEN_PPM / TH_U_EXP * L1033, false, true, false);

           Expression expressionPpmTh_RM = buildExpression(TH_CONCEN_PPM_RM,
                  U_CONCEN_PPM_RM * TH_U_EXP_RM / L1033, true, false, false);

           Expression expressionPpmURM = buildExpression(U_CONCEN_PPM_RM,
                   TH_CONCEN_PPM_RM / TH_U_EXP_RM * L1033, true, false, false);
            }
```
note also that the third expression just above has the correct reference to TH_U_EXP_RM referred to as a bug in this issue.

It would appear that a likely solution is that the second expression in the Thorium case should follow the pattern of the second expression in the Uranium case.

**Comment by sbodorkos** [comment]
Yes - your logic is correct, and rigorous mirroring is what we need to do.

Note that your 'likely solution' would **nominally** populate an expression-name like [correction-independent] Th_Concen_RM, which would **theoretically** be represented in the RM Report Table as "Th (ppm)", by analogy with how "U (ppm)" is (invariantly) defined and reported right now.

But I don't think we should be Report-Tabling a newly-defined and additional "Th (ppm)"; instead, we should simply populate the existing "204corr Th (ppm)", "207corr Th (ppm)", and (in Perm 1) "208corr Th (ppm)" with the same (correction-independent) "Th (ppm)" value that is calculated by the amended expression following your suggestion.

But **note** that this extended behaviour for Th would also need to be rolled out for U, **in its entirety**. That would mean that Report-Tables ought to have, instead of a single "U (ppm)" column, three columns "204corr U (ppm)", "207corr U (ppm)", and (for Perm1) "208corr U (ppm)". 

Now, the row-by-row sets of Xcorr U (ppm) values will continue to be equivalent and identical for U-centric Tasks of any Perm, as well as Perm1 and Perm3 Tasks that are Th-centric. But it will allow us to deal properly with the currently unaccounted cases, corresponding to Perm2 and Perm4 Tasks where the CRM is defined in terms of Th. In these cases, there exist real and distinct values for 204corr U (ppm) and 207corr U (ppm) on a row-by-row basis, but our current Report Tables do not recognise this possibility, which makes me suspect that the underlying code does not yet allow those calculations to occur. Rigorous mirroring will fix that.

**Comment by sbodorkos** [comment]
I have teased out and tabulated all the permutations of Perm and Primary Element, and their influence on how the various U (ppm) and Th (ppm) values are actually calculated:

![image](https://user-images.githubusercontent.com/17473203/56248055-24f28c80-60ea-11e9-9474-47d44acceb6b.png)

**Yellow** denotes the Primary Element. Its value is always calculated using a PARENT_ELEMENT-based expression (which you have already implemented for U where appropriate, and which your proposal will implement for Th where appropriate). PARENT_ELEMENT-based expressions are always independent of index isotope.

**Orange** denotes the Secondary Element in Perm1 and Perm3 (i.e. when DirectAltPD = FALSE). Its value is always calculated as a function of Primary Element together with EqNum -3 (i.e. "expression for 232Th/238U"). Again, the values calculated are independent of index isotope.

**Green** denotes Secondary Element = Th in Perm2 and Perm4 (i.e. when DirectAltPD = TRUE). Its value is always calculated as a function of the two Calibration Constants defined by EqNum -1 and EqNum -2, and therefore depends on the (single) index isotope used to correct both calibration constants. Thus our arithmetic dictates that V4 and V7 should have different numeric values. Note that Squid3 performs these calculations **correctly** in Perm2 and Perm4 **when Secondary Element = Th**. 

**Red** denotes Secondary Element = U in Perm2 and Perm4 (i.e. when DirectAltPD = TRUE), which are the "unaccounted" cases I described above. Its value **should** be calculated as a function of the two Calibration Constants defined by EqNum -1 and EqNum -2, and should therefore depend on the (single) index isotope used to correct both calibration constants. Thus our arithmetic dictates that V4 and V7 should have different numeric values. At present, Squid3 seems not to do this, and it should be implemented by mirroring the Th-based calculations (**green**), which seem to work correctly.
